src/rt/lifetime.d

```diff
@@ -551,20 +655,32 @@ extern(C) void _d_arrayshrinkfit(const TypeInfo ti, void[] arr) /+nothrow+/
     // note, we do not care about shared.  We are setting the length no matter
     // what, so no lock is required.
     debug(PRINTF) printf("_d_arrayshrinkfit, elemsize = %d, arr.ptr = x%x arr.length = %d\n", ti.next.tsize, arr.ptr, arr.length);
-    auto size = ti.next.tsize;                  // array element size
+    auto tinext = unqualify(ti.next);
+    auto size = tinext.tsize;                  // array element size
     auto cursize = arr.length * size;
-    auto   bic = __getBlkInfo(arr.ptr);
-    auto   info = bic ? *bic : GC.query(arr.ptr);
+    auto bic = __getBlkInfo(arr.ptr);
+    auto info = bic ? *bic : GC.query(arr.ptr);
     if(info.base && (info.attr & BlkAttr.APPENDABLE))
     {
-        if(info.size >= PAGESIZE)
-            // remove prefix from the current stored size
-            cursize -= LARGEPREFIX;
+        auto newsize = (arr.ptr - __arrayStart(info)) + cursize;
+
         debug(PRINTF) printf("setting allocated size to %d\n", (arr.ptr - info.base) + cursize);
 
+        // destroy structs that become unused memory when array size is shrinked
+        if (typeid(tinext) is typeid(TypeInfo_Struct)) // avoid a complete dynamic type cast
+        {
+            auto sti = cast(TypeInfo_Struct)cast(void*)tinext;
+            if (sti.xdtor)
+            {
+                auto oldsize = __arrayAllocLength(info, tinext);
+                if (oldsize > cursize)
+                    finalize_array(arr.ptr + cursize, oldsize - cursize, sti);
+            }
+        }
         // Note: Since we "assume" the append is safe, it means it is not shared.
         // Since it is not shared, we also know it won't throw (no lock).
-        __setArrayAllocLength(info, (arr.ptr - info.base) + cursize, false);
+        if (!__setArrayAllocLength(info, newsize, false, tinext))
+            onInvalidMemoryOperationError();
```

Comment:
@rainers  This change is forcing us to lie about the throwability of `_d_arrayshinkfit` to `assumeSafeAppend`.   See https://github.com/dlang/druntime/pull/2647/files#r295209606

Also, it seems to be in conflict with the comment right above it says "it won't throw".  I know it's been an awfully long time ago, but is it possible to remove this in order to make `_d_arrayshrinkfit` `nothrow` so we don't have to do such hacks.

Reply:
onInvalidMemoryOperationError is `nothrow` as it throws an Error.

Reply:
Ah, the problem is `finalize_array`. I don't see how this can be guaranteed to be nothrow, as it calls arbitrary destructors. I don't think _d_arrayshrinkfit can be assumed nothrow if the whole thing isn't made a template.

Reply:
Sorry, I don't understand.  How would making it a template help?

Reply:
Templated `_d_arrayshrinkfit` and `finalize_array` could be inferred to be `nothrow` if the destructor of the struct is `nothrow` (if it is an array of structs to begin with).